File: SI/2018-1/LP1/Prática 1/Java/Conta.java
abstract class Operacao {
	public abstract double calcular(double v1, double v2);
}

class Soma extends Operacao {
	public double calcular(double v1, double v2) {
			return v1 + v2;
		}
}

class Subtracao extends Operacao {
	public double calcular(double v1, double v2) {
			return v1 - v2;
		}
}

class Multiplicacao extends Operacao {
	public double calcular(double v1, double v2) {
			return v1 * v2;
		}
}

class Conta {
	public void calculo(Operacao op, double v1, double v2) {
			System.out.println("Resultado: " + op.calcular(v1, v2));
	}
	public static void main(String args[]) { // A main é estática pois é assim que o java identifica ele para ser executada, um método global
		Conta cont = new Conta();
		cont.calculo(new Soma(), 8.0, 4.0);
		cont.calculo(new Subtracao(), 8.0, 4.0);
		cont.calculo(new Multiplicacao(), 8.0, 4.0);
	}
}

File: SI/2018-1/LP1/Prática 1/C++/Conta.cpp
#include <iostream>

using namespace std;

class Operacao {
	public:
		virtual double calcular(double v1, double v2) = 0;
};

class Soma : public Operacao {
	public:
		double calcular(double v1, double v2) {
			return v1 + v2;
		}
};

class Subtracao : public Operacao {
	public:
		double calcular(double v1, double v2) {
			return v1 - v2;
		}
};

class Multiplicacao : public Operacao {
	public:
		double calcular(double v1, double v2) {
			return v1 * v2;
		}
};

class Conta {
	public:
		void calculo(Operacao *op, double v1, double v2) {
			cout << "Resultado: "  << op->calcular(v1, v2) << endl;
		}
};

int main() {
	Conta cont;
	cont.calculo(new Soma(), 8.0, 4.0);
	cont.calculo(new Subtracao(), 8.0, 4.0);
	cont.calculo(new Multiplicacao(), 8.0, 4.0);
	return 0;
}